### DISCUSSION TITLE
What needs to be built first?

### DISCUSSION THREAD

**Discussion by ghost** [tui-cs/Terminal.Gui#1376] [Q&A] [answered]
Sorry for dumb questions, but I'm not a visual studio developer.  I've just compiled existing code at the office, made minor changes, and rebuilt standalone apps.

Regarding UCatalog, I tried first building UCatalog.csproj in Visual Studio 2019, but it says 

     Error	CS0006	Metadata file '..\GUI_CS\Terminal.Gui\bin\Debug\net5.0\ref\Terminal.Gui.dll' could not be found

Then I tried Terminal.Gui.csproj, but it says the same thing.  So I tried opening Terminal.sln in the root and building, but they all return the same missing dll error.

The only instance of that dll Terminal.Gui.dll is from 05/17/2021, and is in ../GUI_CS\ReactiveExample\bin\Debug\netcoreapp3.1

Really sorry if this should all be obvious, but I just want to test the tree navigation.  Is it possible to build UCatalog and test it, without learning all of its dependencies?  Or at least a simple way to point the UCatalog project to the netcoreapp3.1 folder?

(Edit:  I was able to build ..\GUI_CS\StandaloneExample and it works fine, but there doesn't seem to be a treeview in that particular project.)

**Comment by tznind** [comment]
Youl should be able to just `cd` to the `UICatalog` folder and type `dotnet run` which should build and run.  If it doesn't work then check that there are no changes in your local repo vs the one here.  If it still doesn't work it's possible you don't have the dotnet runtime installed properly or something.  If you post the build output from `dotnet run` we can take a look.

```bash
cd .\Repos\gui.cs\UICatalog\
dotnet run
```

**Comment by BDisp** [comment, marked as answer]
> Regarding UCatalog, I tried first building UCatalog.csproj in Visual Studio 2019, but it says

It seems you are using `Windows OS`.

> The only instance of that dll Terminal.Gui.dll is from 05/17/2021, and is in ../GUI_CS\ReactiveExample\bin\Debug\netcoreapp3.1

That because it uses the NuGet package which run on .Net3.1.

> (Edit: I was able to build ..\GUI_CS\StandaloneExample and it works fine, but there doesn't seem to be a treeview in that particular project.)

The same here, using the NuGet package .Net3.1.

Maybe you don't have the .Net5.0 installed and if it is the case you must install it to be able to compile the solution.

**Reply by tznind** [reply]
Here is the dotnet 5 link https://dotnet.microsoft.com/download/dotnet/5.0

Just download the latest SDK from that site e.g. 5.0.302

**Reply by ghost** [reply]
Hi BDisp and tznind,

I was indeed missing 5.0.  I went ahead and installed that, and got a little further.  Here's the output, so far:

dotnet run <ENTER>

Welcome to .NET 5.0!
---------------------
SDK Version: 5.0.302

Telemetry
---------
The .NET tools collect usage data in order to help us improve your experience. It is collected by Microsoft and shared with the community. You can opt-out of telemetry by setting the DOTNET_CLI_TELEMETRY_OPTOUT environment variable to '1' or 'true' using your favorite shell.

Read more about .NET CLI Tools telemetry: https://aka.ms/dotnet-cli-telemetry

----------------
Installed an ASP.NET Core HTTPS development certificate.
To trust the certificate run 'dotnet dev-certs https --trust' (Windows and macOS only).
Learn about HTTPS: https://aka.ms/dotnet-https
----------------
Write your first app: https://aka.ms/dotnet-hello-world
Find out what's new: https://aka.ms/dotnet-whats-new
Explore documentation: https://aka.ms/dotnet-docs
Report issues and find source on GitHub: https://github.com/dotnet/core
Use 'dotnet --help' to see available commands or visit: https://aka.ms/dotnet-cli
--------------------------------------------------------------------------------------
../.nuget\packages\minver\2.5.0\build\MinVer.targets(39,5): error MSB3073: The command "dotnet ../..nuget\packages\minver\2.5.0\build\../minver/MinVer.dll" "../GUI_CS\Terminal.Gui" --auto-increment "" --build-metadata "" --default-pre-release-phase "" --minimum-major-minor "" --tag-prefix "v" --verbosity "" --version-override """ exited with code -532462766. [../GUI_CS\Terminal.Gui\Terminal.Gui.csproj]

The build failed. Fix the build errors and run again.

**Reply by tznind** [reply]
Ok so the path listed there is to a nuget package for a package called `MinVer`.  It is on nuget (https://www.nuget.org/packages/MinVer/2.5.0)

The `MinVer.targets` file is:

```xml
<Project xmlns="http://schemas.microsoft.com/developer/msbuild/2003">

  <PropertyGroup>
    <MSBuildAllProjects Condition="'$(MSBuildAssemblyVersion)' == '' Or '$(MSBuildAssemblyVersion)' &lt; '16.0'">$(MSBuildAllProjects);$(MSBuildThisFileFullPath)</MSBuildAllProjects>
    <GetPackageVersionDependsOn>$(GetPackageVersionDependsOn);MinVer</GetPackageVersionDependsOn>
  </PropertyGroup>

  <PropertyGroup>
    <MinVerDetailed>low</MinVerDetailed>
    <MinVerDetailed Condition="'$(MinVerVerbosity)' == 'detailed' Or '$(MinVerVerbosity)' == 'd' Or '$(MinVerVerbosity)' == 'diagnostic' Or '$(MinVerVerbosity)' == 'diag'">high</MinVerDetailed>
    <NoWarn>$(NoWarn);NU5105</NoWarn>
  </PropertyGroup>

  <Target Name="MinVer" BeforeTargets="BeforeCompile;GetAssemblyVersion;CoreCompile;GenerateNuspec" Condition="'$(DesignTimeBuild)' != 'true' AND '$(MinVerSkip)' != 'true'">
    <Error Condition="'$(UsingMicrosoftNETSdk)' != 'true'" Code="MINVER0001" Text="MinVer only works in SDK-style projects." />
    <Message Importance="$(MinVerDetailed)" Text="MinVer: [input] MSBuildProjectDirectory=$(MSBuildProjectDirectory)" />
    <Message Importance="$(MinVerDetailed)" Text="MinVer: [input] MinVerAutoIncrement=$(MinVerAutoIncrement)" />
    <Message Importance="$(MinVerDetailed)" Text="MinVer: [input] MinVerBuildMetadata=$(MinVerBuildMetadata)" />
    <Message Importance="$(MinVerDetailed)" Text="MinVer: [input] MinVerDefaultPreReleasePhase=$(MinVerDefaultPreReleasePhase)" />
    <Message Importance="$(MinVerDetailed)" Text="MinVer: [input] MinVerMinimumMajorMinor=$(MinVerMinimumMajorMinor)" />
    <Message Importance="$(MinVerDetailed)" Text="MinVer: [input] MinVerTagPrefix=$(MinVerTagPrefix)" />
    <Message Importance="$(MinVerDetailed)" Text="MinVer: [input] MinVerVerbosity=$(MinVerVerbosity)" />
    <Message Importance="$(MinVerDetailed)" Text="MinVer: [input] MinVerVersionOverride=$(MinVerVersionOverride)" />
    <ItemGroup>
      <MinVerInputs Remove="@(MinVerInputs)" />
      <MinVerConsoleOutput Remove="@(MinVerConsoleOutput)" />
      <MinVerOutputVersion Remove="@(MinVerOutputVersion)" />
    </ItemGroup>
    <ItemGroup>
      <MinVerInputs Include="&quot;$(MSBuildProjectDirectory)&quot;" />
      <MinVerInputs Include="--auto-increment &quot;$(MinVerAutoIncrement)&quot;" />
      <MinVerInputs Include="--build-metadata &quot;$(MinVerBuildMetadata)&quot;" />
      <MinVerInputs Include="--default-pre-release-phase &quot;$(MinVerDefaultPreReleasePhase)&quot;" />
      <MinVerInputs Include="--minimum-major-minor &quot;$(MinVerMinimumMajorMinor)&quot;" />
      <MinVerInputs Include="--tag-prefix &quot;$(MinVerTagPrefix)&quot;" />
      <MinVerInputs Include="--verbosity &quot;$(MinVerVerbosity)&quot;" />
      <MinVerInputs Include="--version-override &quot;$(MinVerVersionOverride)&quot;" />
    </ItemGroup>
    <Exec Command="dotnet &quot;$(MSBuildThisFileDirectory)../minver/MinVer.dll&quot; @(MinVerInputs->'%(Identity)', ' ')" ConsoleToMSBuild="true" StandardOutputImportance="Low" >
      <Output TaskParameter="ConsoleOutput" ItemName="MinVerConsoleOutput" />
    </Exec>
    <ItemGroup>
      <MinVerOutputVersion Include="@(MinVerConsoleOutput)" Condition="'$([System.String]::new(`%(Identity)`).StartsWith(`MinVer:`))' != 'true'" />
    </ItemGroup>
    <PropertyGroup>
      <MinVerVersion>@(MinVerOutputVersion)</MinVerVersion>
      <MinVerMajor>$(MinVerVersion.Split(`.`)[0])</MinVerMajor>
      <MinVerMinor>$(MinVerVersion.Split(`.`)[1])</MinVerMinor>
      <MinVerPatch>$(MinVerVersion.Split(`.`)[2].Split(`-`)[0].Split(`+`)[0])</MinVerPatch>
      <MinVerPreRelease></MinVerPreRelease>
      <MinVerPreRelease Condition="$(MinVerVersion.Split(`+`)[0].Contains(`-`))">$(MinVerVersion.Split(`+`)[0].Split(`-`, 2)[1])</MinVerPreRelease>
      <MinVerBuildMetadata></MinVerBuildMetadata>
      <MinVerBuildMetadata Condition="$(MinVerVersion.Contains(`+`))">$(MinVerVersion.Split(`+`, 2)[1])</MinVerBuildMetadata>
      <AssemblyVersion>$(MinVerMajor).0.0.0</AssemblyVersion>
      <FileVersion>$(MinVerMajor).$(MinVerMinor).$(MinVerPatch).0</FileVersion>
      <PackageVersion>$(MinVerVersion)</PackageVersion>
      <Version>$(MinVerVersion)</Version>
    </PropertyGroup>
    <Message Importance="$(MinVerDetailed)" Text="MinVer: [output] MinVerVersion=$(MinVerVersion)" />
    <Message Importance="$(MinVerDetailed)" Text="MinVer: [output] MinVerMajor=$(MinVerMajor)" />
    <Message Importance="$(MinVerDetailed)" Text="MinVer: [output] MinVerMinor=$(MinVerMinor)" />
    <Message Importance="$(MinVerDetailed)" Text="MinVer: [output] MinVerPatch=$(MinVerPatch)" />
    <Message Importance="$(MinVerDetailed)" Text="MinVer: [output] MinVerPreRelease=$(MinVerPreRelease)" />
    <Message Importance="$(MinVerDetailed)" Text="MinVer: [output] MinVerBuildMetadata=$(MinVerBuildMetadata)" />
    <Message Importance="$(MinVerDetailed)" Text="MinVer: [output] AssemblyVersion=$(AssemblyVersion)" />
    <Message Importance="$(MinVerDetailed)" Text="MinVer: [output] FileVersion=$(FileVersion)" />
    <Message Importance="$(MinVerDetailed)" Text="MinVer: [output] PackageVersion=$(PackageVersion)" />
    <Message Importance="$(MinVerDetailed)" Text="MinVer: [output] Version=$(Version)" />
  </Target>

</Project>
```

With the bad line (39) is:

```xml
<Exec Command="dotnet &quot;$(MSBuildThisFileDirectory)../minver/MinVer.dll&quot; @(MinVerInputs->'%(Identity)', ' ')" ConsoleToMSBuild="true" StandardOutputImportance="Low" >
      <Output TaskParameter="ConsoleOutput" ItemName="MinVerConsoleOutput" />
    </Exec>
```

Looks like MinVer is responsible for this not working for you, maybe something about your environment.  Are you definetly running `dotnet run` from within the `UICatalog` folder of the repo and not the root? (e.g. by typing "cd UICatalog" before "dotnet build")

**Reply by BDisp** [reply]
Sometimes it is needed to run `dotnet clean` first.
But if you are using the `Visual Studio 2019`, you can build and run from the `IDE`.

**Reply by tznind** [reply]
I also note that [MinVer includes a requirement](https://github.com/adamralph/minver#prerequisites) for `git`

> Prerequisites
>    .NET Core SDK 2.1 or later
>   Git

You can check if git is installed by running:

`git status`.

If you get 'unknown command' or something from that then you might want to install git https://git-scm.com/downloads.

Incidentally that is the accepted answer for this post on stack overflow which looks like the same issue you have: https://stackoverflow.com/questions/60232802/unable-to-build-identityserver4-on-windows-10-machine

**Reply by tznind** [reply]
Can confirm that not having `git` installed results in not being able to build/run and it causes this exact error message (see https://github.com/migueldeicaza/gui.cs/issues/1377)

**Comment by ghost** [comment]
tznind and BDisp, thanks for tracking down the source of the error.

I'll look at the code for now.  During the week, I'll open a laptop that has git installed.

**Reply by tznind** [reply]
Thanks for persevering, if you add the following to Terminal.Gui/Terminal.Gui.csproj then you wont need git with `dotnet run`:

```xml
  <PropertyGroup>
      <MinVerSkip Condition="'$(Configuration)' == 'Debug'">true</MinVerSkip>
  </PropertyGroup>
```

See this PR for more details: https://github.com/migueldeicaza/gui.cs/pull/1378

**Reply by ghost** [reply]
That did the trick, tznind.  The .dll was generated without errors.

Thanks.

Edit:  

No way.  Miguel threw in a text mode CSV editor?  Holy sht.

Good stuff.  It'll be fun playing with it this week, thanks everyone.
